initial commit of tfidf transformer

pazzo83 opened the pull request:
Initial commit - a few notes

- I wasn't sure on the input/output types, but what how I'm picturing this working is one would be able to pass to the transformer either a vector of `String`s or a vector of `StringDocument`s from the TextAnalysis package.  Then the various fit functions apply the appropriate conversions, create n_grams (if necessary), etc.  If you think this isn't the right approach, I am definitely open to other ways.
- I've introduced some checks on parameter settings, but I am not 100% sure where those should live.

This is definitely in a more raw state than the TSVD code - so please let me know what you think should be changed/removed/added.  Thanks!

ablaom commented:
Done.

storopoli commented on Project.toml:
```diff
@@ -1,24 +1,25 @@
 [compat]
 MLJModelInterface = "1.1.1"
-ScientificTypesBase = "1"
+ScientificTypesBase = "2.2.0"
```
Wouldn't it better to:

```suggestion
ScientificTypesBase = "2.2"
```

storopoli replied:
Do we also need ScientificTypesBase 2.2.2 as well?
@ablaom

storopoli commented on Project.toml:
```diff
@@ -1,24 +1,25 @@
 TextAnalysis = "a2db99b7-8b79-58f8-94bf-bbc811eef33d"
 
 [compat]
 MLJModelInterface = "1.1.1"
```
Wouldn't it better to:
```suggestion
MLJModelInterface = "1.1"
```

storopoli commented on Project.toml:
```diff
@@ -1,24 +1,25 @@
 MLJModelInterface = "1.1.1"
-ScientificTypesBase = "1"
+ScientificTypesBase = "2.2.0"
+ScientificTypes = "2.2.0"
```
Wouldn't it better to:
```suggestion
ScientificTypes = "2.2"
```

ablaom replied:
I think we actually need ScientificTypes 2.2.2 to get [this fix](https://github.com/JuliaAI/ScientificTypes.jl/pull/158)

storopoli replied:
Ok, changed.

storopoli commented on Project.toml:
```diff
@@ -1,24 +1,25 @@
-ScientificTypesBase = "1"
+ScientificTypesBase = "2.2.0"
+ScientificTypes = "2.2.0"
 TextAnalysis = "0.7.3"
```
Wouldn't it better to:
```suggestion
TextAnalysis = "0.7"
```

pazzo83 replied:
on this want we need the latest for now (0.7.3)

storopoli commented on src/MLJText.jl:
```diff
@@ -1,115 +1,192 @@
+
+Convert a collection of raw documents to a matrix of TF-IDF features.
+
+"Tf" means term-frequency while "tf-idf" means term-frequency times
```
I prefer the Docstring capitalised as: 
```suggestion
"TF" means term-frequency while "TF-IDF" means term-frequency times
```

storopoli commented on src/MLJText.jl:
```diff
@@ -1,115 +1,192 @@
-A cool classifier that predicts `UnivariateFinite` probability
-distributions. These are distributions for a finite sample space whose
-elements are *labeled*.
+    TfidfTransformer()
```
And the `struct` as CameCase like julia convetions:
```suggestion
    TfIdfTransformer()
```

storopoli replied:
this probably needs to be changed downstream also

pazzo83 replied:
this follows sklearn's representation of the object/class/struct in camel case - I think it's ok for now?  Don't mind changing either though.

storopoli replied:
We better follow the sklearn guidelines. My bad

storopoli reviewed (approved):
Here are some opinions. I think as an initial idea we should be much more concerned with naming and API-user interaction than with performance.

pazzo83 commented:
Ok @ablaom @storopoli  - I made a few additional changes.  I updated the deps - is it ok to have `MLJModelInterface` set to just 1.3?  I'll use whatever the convention is there (figured I'd bump it up to at least that since 1.3.2 is the current version).  I also implemented the switch in orientation of the actual TF-IDF matrix that we discussed.  It does seem to be quite a bit faster - and works just fine in some of the local tests I have.  Looks like it is complaining about the `2.2.2` version for `ScientificTypesBase`.

pazzo83 commented:
Also, you all might notice that we keep failing in the Julia 1.3 check.  I was doing some testing with different Julia versions (thank you Google Colab), and the issue appears to be how I am specifying types in the `build_corpus` method.  As you see, I've defined more or less a type alias for NGrams as such:
```julia
const NGram{N} = NTuple{<:Any,<:AbstractString}
```

Then I use that type alias for multiple dispatch in a function that constructs the proper `TextAnalysis.Corpus` object:

```julia
_convert_bag_of_words(X::Dict{NGram, Int}) = 
    Dict(join(k, " ") => v for (k, v) in X)

build_corpus(X::Vector{Dict{NGram, Int}}) = 
    build_corpus(_convert_bag_of_words.(X))
build_corpus(X::Vector{Dict{S, Int}}) where {S <: AbstractString} = 
    Corpus(NGramDocument.(X))
build_corpus(X) = Corpus(TokenDocument.(X))
```

In versions prior to Julia 1.5, the compiler blows up when it tries to deal with a `Dict` with the `NGram` type alias as a key.  So something like this:
```julia
test_dict2 = Dict{NGram, Int}(("a", "b") => 1, ("c", "d", "e") => 2)
```
causes Julia to crash.  However, starting in 1.5, it does just fine.  Should we bump up the Julia version in `Compat` to 1.5?

ablaom commented:
> I updated the deps - is it ok to have MLJModelInterface set to just 1.3? 
Yep, that's fine.

Re your fail, I'm guessing your have some missing `<:`. So what if you try:

```julia
_convert_bag_of_words(X::Dict{<:NGram, <:Integer}) = 
    Dict(join(k, " ") => v for (k, v) in X)

build_corpus(X::Vector{<:Dict{<:NGram, <:Integer}}) = 
    build_corpus(_convert_bag_of_words.(X))
build_corpus(X::Vector{<:Dict{S, <:Integer}}) where {S <: AbstractString} = 
    Corpus(NGramDocument.(X))
build_corpus(X) = Corpus(TokenDocument.(X))
```

ablaom commented:
Sounds reasonable to me. PR welcome :)

ablaom commented on Project.toml:
```diff
@@ -15,7 +15,7 @@ MLJModelInterface = "1.3"
 ScientificTypes = "2.2.2"
 TextAnalysis = "0.7.3"
-julia = "1.3"
+julia = "1.5"
```
Are you still getting a fail? Can you please re-instate 1.3 so I can investigate.  This sounds fishy. If 1.3 is failing it should only be because we are using a newer julia feature.

pazzo83 replied:
it's still failing - CI is still running 1.3 because I see now you have to update the proper file in `.github/workflows`.  I can confirm that the code I posted above (`test_dict2 = Dict{NGram, Int}(("a", "b") => 1, ("c", "d", "e") => 2)`) fails in Julia 1.3.  Whatever issue is causing Julia to crash got fixed/changed sometime between Julia 1.4 and Julia 1.5 (or perhaps was fixed in Julia 1.5 which is the first version I tried where it did not fail).

ablaom replied:
Thanks! I'll look into the this today.

ablaom commented on test/runtests.jl:
```diff
@@ -1,24 +1,63 @@
+        "in" => 1,
+        "the cat" => 1
+    )
+    bag = Dict{MLJText.NGram, Int}(Tuple(String.(split(k))) => v for (k, v) in bag_of_words)
```
Looks like I can now remove the type params.

ablaom commented on test/runtests.jl:
```diff
@@ -1,24 +1,63 @@
+        "in" => 1,
+        "the cat" => 1
+    )
+    bag = Dict{MLJText.NGram, Int}(Tuple(String.(split(k))) => v for (k, v) in bag_of_words)
```
```suggestion
    bag = Dict(Tuple(String.(split(k))) => v for (k, v) in bag_of_words)
```

With this change, I get tests passing on Julia 1.3. This is a reasonable test for correctness of your algorithm, no?  I don't see we need to limit the user to Julia 1.3, as the fail does indeed look like a Julia bug rather than something wrong with with the implementation. 

Thanks for diagnosing the issue. I imagine this was a bit of pain 😥 . No-one wants to see "unreachable..."!

ablaom replied:
@pazzo83

pazzo83 replied:
interesting - it looks like with the change to the methods where I'm trying to build the corpus that you suggested (the `build_corpus` methods), the type that is resolved by simply having `Dict(Tuple(String.(split(k))) => v for (k, v) in bag_of_words)` is able to be routed to the proper method.  Thanks!!
